### PR TITLE
Introduce `Duration` type to keep apart nominal vs exact durations.

### PR DESCRIPTION
**Pull request by minichma**
This PR addresses shortcomings of the lib regarding the different requirements of nominal vs exact duration as specified by RFC 5545. It introduces the new `Duration` type that replaces `TimeSpan` in most places. In contrast to the latter, `Duration` stores the individual duration parts (weeks, days, hours, minutes, seconds) individually. This is required because according to the RFC weeks and days are considered nominal values while hours, minutes, seconds are considered exact (see https://www.rfc-editor.org/rfc/rfc5545#section-3.3.6, https://www.rfc-editor.org/rfc/rfc5545#section-3.8.5.3).

Most relevant changes:

* `CalendarEvent.Duration` et al and `Period.Duration` changed from `TimeSpan` to `Duration`
* `CalendarEvent.GetTimeSpanToAddToPeriodStartTime` replaced by `CalendarEvent.GetEffectiveDuration`, which takes care of keeping apart nominal vs. exact durations. I.e. if the duration is specified via DTEND, then the returned duration only uses time-related fields, so it is considered exact. If specified via `DURATION`, the duration is returned as specified by the caller.
* `CalDateTime.Add(Duration)` and `.Subtract()` handle the different parts of `Duration` conformant to the RFC.

### CODE REVIEW

**Comment by axunonb** [comment]
I'll be waiting until this PR gets merged before submitting a PR for `EXDATE` and `RDATE` fixes.

**Comment by minichma** [comment]
> I'll be waiting until this PR gets merged before submitting a PR for `EXDATE` and `RDATE` fixes.

@axunonb I think the PR should be ready for review, would be great if you'd find some time for it. It's quite some code that has changed, but all in all not too complex, once the basics regarding nominal vs. exact are understood. Probably it makes sense to review the code commit by commit. I tried to keep the commits reasonably separated, but probably not perfectly everywhere. There are certainly some smaller issues around nonexistent times remaining as outlined in #681.

File: Ical.Net/Serialization/DataTypes/DurationSerializer.cs
Comment (axunonb) [inline Ical.Net/Serialization/DataTypes/DurationSerializer.cs:22]: Maybe just
```cs 
public override string SerializeToString(object obj)
    => !(obj is Duration duration) ? null : SerializeToString(duration);
```

File: Ical.Net/Serialization/DataTypes/DurationSerializer.cs
Comment (axunonb) [inline Ical.Net/Serialization/DataTypes/DurationSerializer.cs]: better use StringBuilder.Append(char) instead of string?

File: Ical.Net/DataTypes/Duration.cs
Comment (axunonb) [inline Ical.Net/DataTypes/Duration.cs:196]: Would `Duration? x = null; if (x == null)` be more clear/simple than than `Duration x = new Duration(); if (x.IsEmpty)`? Then we could disallow empty.

File: Ical.Net/DataTypes/IDateTime.cs
Comment (axunonb) [inline Ical.Net/DataTypes/IDateTime.cs]: A xmldoc hint for SubtractExact vs Subtract similar like in static Duration FromTimeSpan(TimeSpan t) might be useful.

"from to the value" - "to" is too much?

File: Ical.Net/DataTypes/Duration.cs
Comment (axunonb) [inline Ical.Net/DataTypes/Duration.cs:214]: or
```cs
v switch
{
    null => null,
    >= 0 => 1,
    _ => -1
};
```

**Comment by axunonb** [comment]
The plan for a next PR re `Period`was:
- Change `public Perdiod()` to internal (should not be part of the public API, see #676). Change`Activator.CreateInstance(type, true)` in `CalendarDataType.Copy`.
- Timezone of StartTime and EndTime (if set) must be the same (as needed in the iCalendar context anyway)
That's necessary for EXDATE and RDATE.

File: Ical.Net/DataTypes/Period.cs
Comment (axunonb) [inline Ical.Net/DataTypes/Period.cs]: Isn't this redundant now as these properties already represent the original values?

File: Ical.Net/DataTypes/Period.cs
Comment (axunonb) [inline Ical.Net/DataTypes/Period.cs:126]: Like it!

File: Ical.Net/DataTypes/Period.cs
Comment (axunonb) [inline Ical.Net/DataTypes/Period.cs:140]: Nice, too!

File: Ical.Net/DataTypes/Period.cs
Comment (axunonb) [inline Ical.Net/DataTypes/Period.cs:179]: much better

File: Ical.Net/DataTypes/CalDateTime.cs
Comment (axunonb) [inline Ical.Net/DataTypes/CalDateTime.cs]: You asked whether `otherTzId` should be `string?`, and the answer was no.
But, well, sometimes it would be convenient to easily change a zoned date/time to Floating this way., WDYT - keep or change?

File: Ical.Net/CalendarComponents/Todo.cs
Comment (axunonb) [inline Ical.Net/CalendarComponents/Todo.cs:194]: One more time...

File: Ical.Net/DataTypes/CalDateTime.cs
Comment (axunonb) [inline Ical.Net/DataTypes/CalDateTime.cs:492]: A xmldoc hint for SubtractExact vs Subtract similar like in static Duration FromTimeSpan(TimeSpan t) might be useful.

File: Ical.Net/DataTypes/CalDateTime.cs
Comment (axunonb) [inline Ical.Net/DataTypes/CalDateTime.cs:500]: good

File: Ical.Net.Tests/RecurrenceTests.cs
Comment (axunonb) [inline Ical.Net.Tests/RecurrenceTests.cs:138]: Shouldn't we reduce the size of the CalDateTime lists in RecurrenceTests to the relevant entries (which is much smaller)?

**Review by axunonb — COMMENTED** [review]
Excellent.
See some minor remarks.
{Edit] Just forgot to mention the nullability warnings like Possible null reference argument for parameter 'dt' in 'Duration IDateTime.Subtract(IDateTime dt)' et al

File: Ical.Net/DataTypes/Duration.cs
Comment (minichma) [reply Ical.Net/DataTypes/Duration.cs:196]: > Would Duration? x = null; if (x == null) be more clear/simple than than Duration x = new Duration(); if (x.IsEmpty)?

Basically yes, but we don't use `new Duration()` like that anyways. The only place we use it is to initialize `public static Duration Zero { get; } = new Duration();`. Not sure whether it would be desirable to disallow an empty Duration. Would rather say we should allow it. TBD.

File: Ical.Net/Serialization/DataTypes/DurationSerializer.cs
Comment (minichma) [reply Ical.Net/Serialization/DataTypes/DurationSerializer.cs:22]: Certainly, thanks. This class is based on a copy of the previous `TimeSpanSerializer`,  could easily see some leftovers from there.

File: Ical.Net/DataTypes/Duration.cs
Comment (axunonb) [reply Ical.Net/DataTypes/Duration.cs:196]: Oh yes, right, then `static Duration Zero { get; } = new Duration();` returns a duration that `IsEmpty`, not `IsZero`?

File: Ical.Net/DataTypes/Duration.cs
Comment (minichma) [reply Ical.Net/DataTypes/Duration.cs:214]: Nice, haven't been aware of that syntax.

File: Ical.Net/DataTypes/Period.cs
Comment (minichma) [reply Ical.Net/DataTypes/Period.cs]: yes, absolutely

File: Ical.Net/DataTypes/CalDateTime.cs
Comment (minichma) [reply Ical.Net/DataTypes/CalDateTime.cs]: As TzId `null` represents floating time everywhere throughout the class, it would be logical to allow `null` here. Otherwise we should introduce a method `ToFloating()`, but allowing null here keeps the API more straight forward. So I'd say, `null` should be allowed.

File: Ical.Net/CalendarComponents/Todo.cs
Comment (minichma) [reply Ical.Net/CalendarComponents/Todo.cs:194]: No more `Extrapolate` for now :-)

File: Ical.Net/DataTypes/CalDateTime.cs
Comment (minichma) [reply Ical.Net/DataTypes/CalDateTime.cs:500]: Yeah, I also tried throwing if `IsFloating != dt.IsFloating`, but then all kinds of tests fail. Obviously there are quite some cases where `DTSTART` has a tzid set but `DTEND` hasn't, which isn't allowed by the RFC though. We should double-check whether we still want to support this case (probably yes, if its used in the wild). If so, we should probably deal with the case in a different place, probably during deserialization, and be more strict here, because the current implementation is somewhat unclear regarding that case.

File: Ical.Net.Tests/RecurrenceTests.cs
Comment (minichma) [reply Ical.Net.Tests/RecurrenceTests.cs:138]: Don't know enough about the individual tests to comment on that, but sounds like quite some work. :-) Certainly not in scope of this PR though.

File: Ical.Net/DataTypes/Duration.cs
Comment (minichma) [reply Ical.Net/DataTypes/Duration.cs:196]: Actually it is both, empty and zero.

I.e. `new Duration()` is both, empty and zero,
`new Duration(0)` is zero but not empty.

Both are `internal`, as I'm aware that it could be confusing.

File: Ical.Net/DataTypes/CalDateTime.cs
Comment (minichma) [reply Ical.Net/DataTypes/CalDateTime.cs]: Added support for null as I'm on it anyways.

**Comment by minichma** [comment]
@axunonb Thanks for the review! Tried to improve.

**Review by axunonb — APPROVED** [review]
Great job, thanks a lot

**Pull request merged**
